pkg/backup/compaction_dist.go

```diff
@@ -0,0 +1,334 @@
+		nil, /* txn */
+		nil, /* clockUpdater */
+		execCtx.ExtendedEvalContext().Tracing,
+	)
+	defer recv.Release()
+
+	jobsprofiler.StorePlanDiagram(
+		ctx, execCtx.ExecCfg().DistSQLSrv.Stopper, plan, execCtx.ExecCfg().InternalDB, jobID,
+	)
+
+	evalCtxCopy := execCtx.ExtendedEvalContext().Copy()
+	dsp.Run(ctx, planCtx, nil /* txn */, plan, recv, evalCtxCopy, nil /* finishedSetupFn */)
+	return nil
+}
+
+// createCompactionPlan creates an un-finalized physical plan that will
+// distribute spans from a generator across the cluster for compaction.
+func createCompactionPlan(
+	ctx context.Context,
+	execCtx sql.JobExecContext,
+	jobID jobspb.JobID,
+	details jobspb.BackupDetails,
+	manifest *backuppb.BackupManifest,
+	dsp *sql.DistSQLPlanner,
+	genSpan func(ctx context.Context, spanCh chan execinfrapb.RestoreSpanEntry) error,
+	spansToCompact roachpb.Spans,
+	targetSize int64,
+	maxFiles int64,
+) (*sql.PhysicalPlan, *sql.PlanningCtx, error) {
+	numEntries, err := countRestoreSpanEntries(ctx, genSpan)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "counting number of restore span entries")
+	}
+
+	// TODO (kev-cao): Add support for execution locality.
+	planCtx, sqlInstanceIDs, err := dsp.SetupAllNodesPlanningWithOracle(
+		ctx, execCtx.ExtendedEvalContext(), execCtx.ExecCfg(),
+		physicalplan.DefaultReplicaChooser, roachpb.Locality{},
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	plan := planCtx.NewPhysicalPlan()
+	corePlacements, err := createCompactionCorePlacements(
+		ctx, jobID, execCtx.User(), details, manifest.ElidedPrefix, genSpan,
+		spansToCompact, sqlInstanceIDs, targetSize, maxFiles, numEntries,
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+	plan.AddNoInputStage(
+		corePlacements,
+		execinfrapb.PostProcessSpec{},
+		[]*types.T{},
+		execinfrapb.Ordering{},
+		nil, /* finalizeLastStageCb */
+	)
+	return plan, planCtx, nil
+}
+
+// countRestoreSpanEntries counts the number of restore span entries that will be
+// be delivered by the given generator.
+func countRestoreSpanEntries(
+	ctx context.Context,
+	genSpan func(ctx context.Context, spanCh chan execinfrapb.RestoreSpanEntry) error,
+) (int, error) {
+	countSpansCh := make(chan execinfrapb.RestoreSpanEntry, 1000)
+	var numImportSpans int
+	countTasks := []func(ctx context.Context) error{
+		func(ctx context.Context) error {
+			for range countSpansCh {
+				numImportSpans++
+			}
+			return nil
+		},
+		func(ctx context.Context) error {
+			defer close(countSpansCh)
+			return genSpan(ctx, countSpansCh)
+		},
+	}
+	if err := ctxgroup.GoAndWait(ctx, countTasks...); err != nil {
+		return 0, errors.Wrapf(err, "counting number of spans to compact")
+	}
+	return numImportSpans, nil
+}
+
+// createCompactionCorePlacements takes spans from a generator and evenly
+// distributes them across nodes in the cluster, returning the core placements
+// reflecting that distribution.
+//
+// TODO (kev-cao): This currently distributes the spans so that each node gets
```

Comment:
the spans outputted by genSpans are split on size as well?

Reply:
> // Additionally, if targetSize > 0, we can
// merge the current cover span with the previous cover span if the merged set
// of files have a total data size below the target size.

Oh you're right, well that's nice of it